packages/core/src/components/textStream.ts

```diff
@@ -55,61 +55,71 @@ export function setupTextStream(room: Room, topic: string): Observable<TextStrea
-      scan((acc: string, chunk: string) => {
-        return acc + chunk;
-      }, ''),
-    );
-
-    const isTranscription = !!reader.info.attributes?.[segmentAttribute];
-
-    // Subscribe to the stream and update our array when new chunks arrive
-    streamObservable.subscribe((accumulatedText) => {
-      // Find and update the stream in our array
-      const index = textStreams.findIndex(
-        (stream) =>
-          stream.streamInfo.id === reader.info.id ||
-          (isTranscription &&
-            stream.streamInfo.attributes?.[segmentAttribute] ===
-              reader.info.attributes?.[segmentAttribute]),
-      );
-      if (index !== -1) {
-        textStreams[index] = {
-          ...textStreams[index],
-          text: accumulatedText,
-        };
-
-        // Emit the updated array
-        textStreamsSubject.next([...textStreams]);
-      } else {
-        // Handle case where stream ID wasn't found (new stream)
-        textStreams.push({
-          text: accumulatedText,
-          participantInfo,
-          streamInfo: reader.info,
+  // Create shared observable and store in cache
+  const sharedObservable = textStreamsSubject.pipe(
+    tap({
+      subscribe: () => {
+        room.registerTextStreamHandler(topic, async (reader, participantInfo) => {
+          // Create an observable from the reader
+          const streamObservable = from(reader).pipe(
+            scan((acc: string, chunk: string) => {
+              return acc + chunk;
+            }, ''),
+          );
+
+          const isTranscription = !!reader.info.attributes?.[segmentAttribute];
+
+          // Subscribe to the stream and update our array when new chunks arrive
+          streamObservable.subscribe((accumulatedText) => {
+            // Find and update the stream in our array
+            const index = textStreams.findIndex(
+              (stream) =>
+                stream.streamInfo.id === reader.info.id ||
+                (isTranscription &&
+                  stream.streamInfo.attributes?.[segmentAttribute] ===
+                    reader.info.attributes?.[segmentAttribute]),
+            );
+            if (index !== -1) {
+              textStreams[index] = {
+                ...textStreams[index],
+                text: accumulatedText,
+              };
+
+              // Emit the updated array
+              textStreamsSubject.next([...textStreams]);
+            } else {
+              // Handle case where stream ID wasn't found (new stream)
+              textStreams.push({
+                text: accumulatedText,
+                participantInfo,
+                streamInfo: reader.info,
+              });
+
+              // Emit the updated array with the new stream
+              textStreamsSubject.next([...textStreams]);
+            }
+          });
         });
+      },
+      finalize: () => {
+        console.log('finalizing');
+        room.unregisterTextStreamHandler(topic);
```

Comment:
removing the handler on finalize. This works because we also use `share` which does internal refcounting and finalize is only called once the refcount drops to 0